js-examples/addressbook.js

```diff
@@ -0,0 +1,142 @@
+  Person_PhoneNumber_Type
+} = require('./addressbook.capnp.js');
+
+function writePackedMessageToStream(writeStream, message) {
+  const arrayBuffer = message.toPackedArrayBuffer();
+  // Beacause streams can't handle ArrayBuffers
+  const buffer = new Buffer(arrayBuffer);
+  writeStream.write(buffer);
+}
+
+function writeAddressBook(writeStream) {
+  const message = new capnp.Message();
+  const addressBook = message.initRoot(AddressBook);
+  const people = addressBook.initPeople(2);
+
+  const alice = people.get(0);
+  alice.setId(123);
+  alice.setName("Alice");
+  alice.setEmail("alice@example.com");
+  const alicePhones = alice.initPhones(1);
+  alicePhones.get(0).setNumber("555-1212");
+  alicePhones.get(0).setType(Person_PhoneNumber_Type.MOBILE);
+  alice.getEmployment().setSchool("MIT");
+
+  const bob = people.get(1);
+  bob.setId(456);
+  bob.setName("Bob");
+  bob.setEmail("bob@example.com");
+  const bobPhones = bob.initPhones(2);
+  bobPhones.get(0).setNumber("555-4567");
+  bobPhones.get(0).setType(Person_PhoneNumber_Type.HOME);
+  bobPhones.get(1).setNumber("555-7654");
+  bobPhones.get(1).setType(Person_PhoneNumber_Type.WORK);
+  bob.getEmployment().setUnemployed();
+
+  writePackedMessageToStream(writeStream, message);
+}
+
+function readToEndOfStream(readStream) {
+  return new Promise(function(resolve, reject) {
+    let result = new Uint8Array();
+    readStream.on('data', function(data) {
+      const oldLen = result.byteLength;
+      const newResult = new Uint8Array(oldLen + data.byteLength);
+      newResult.set(result);
+      newResult.set(data, oldLen);
+      result = newResult;
+    });
+    readStream.on('end', function() {
+      resolve(result);
+    });
+    readStream.on('error', reject);
+  });
+}
+
+function printAddressBook(readStream) {
+  return readToEndOfStream(readStream).then(function(data) {
+    const message = capnp.Message.fromPackedArrayBuffer(data);
+    const addressBook = message.getRoot(AddressBook);
+    addressBook.getPeople().forEach(function(person) {
+      console.log(person.getName() + ': ' + person.getEmail());
+      person.getPhones().forEach(function(phone) {
+        let phoneTypeString = Person_PhoneNumber_Type[phone.getType()];
+        console.log('  ' + phoneTypeString + ' phone: ' + phone.getNumber());
+      });
+      const employment = person.getEmployment();
+      switch (employment.which()) {
+        case Person_Employment.UNEMPLOYED:
+          console.log('  unemployed');
+          break;
+        case Person_Employment.EMPLOYER:
+          console.log('  employer: ' + employment.getEmployer());
+          break;
+        case Person_Employment.SCHOOL:
+          console.log('  student at: ' + employment.getSchool());
+          break;
+        case Person_Employment.SELF_EMPLOYED:
+          console.log('  self-employed');
+          break;
+      }
+    });
+  });
+}
+
+function main() {
+  // TODO Not sure if we have dynamic schema usage available yet
```

Comment:
A dynamic API is useful for writing generic algorithms that operate on arbitrary Cap'n Proto objects. For example, say you wanted to write a library for converting capnp types to/from JSON. This kind of thing usually requires schema information (maybe you can infer everything you need from native Javascript reflection, but often you can't). So your options are a dynamic API, or offline code generation.

Another use case that's particularly relevant to Javascript: It would be really cool to write a dynamic RPC interface explorer client. The client could connect to an RPC endpoint and dynamically request the interface's schema, then present a UI to the user to explore and manually invoke methods. In this case you don't even know the schemas at compile time, so a dynamic API is probably the only option.

This stuff probably isn't necessary for V1 but I think people will want it eventually.